sw/device/tests/crypto/cryptotest/firmware/ibex_fi.c

```diff
@@ -30,17 +33,184 @@
   ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 ADDI100 \
       ADDI100
 
+// Init x6 = 10000 macro.
+#define INITX6 "li x6, 10000"
+
+// Subi x6 = x6 - 1 macro.
+#define SUBI1 "addi x6, x6, -1\n"
+
+// Load word, addi, sw macro.
+#define LWADDISW1 "lw x5, (%0)\n addi x5, x5, 1\n sw x5, (%0)\n"
+#define LWADDISW10                                                      \
+  LWADDISW1 LWADDISW1 LWADDISW1 LWADDISW1 LWADDISW1 LWADDISW1 LWADDISW1 \
+      LWADDISW1 LWADDISW1 LWADDISW1
+#define LWADDISW100                                                            \
+  LWADDISW10 LWADDISW10 LWADDISW10 LWADDISW10 LWADDISW10 LWADDISW10 LWADDISW10 \
+      LWADDISW10 LWADDISW10 LWADDISW10
+#define LWADDISW1000                                                      \
+  LWADDISW100 LWADDISW100 LWADDISW100 LWADDISW100 LWADDISW100 LWADDISW100 \
+      LWADDISW100 LWADDISW100 LWADDISW100 LWADDISW100
+
+// Load word, subi, sw macro.
+#define LWSUBISW1 "lw x6, (%0)\n addi x6, x6, -1\n sw x6, (%0)\n"
+
 /**
- * Initializes the SCA trigger.
+ * ibex.char.mem_op_loop command handler.
  *
+ * This FI penetration tests executes the following instructions:
+ * - Add 100 NOPs to delay the trigger
+ * - 10000 iterations with a for loop:
+ *  - Load loop_counter1 value into x5: lw x5, (&loop_counter1)
+ *  - Increment loop counter1: addi x5, x5, 1
+ *  - Store loop counter1 back to loop_counter1: sw x5, (&loop_counter1)
+ *  - Load loop_counter2 value into x6: lw x6, (&loop_counter2)
+ *  - Decrement loop counter2: addi x6, x6, -1
+ *  - Store loop counter2 back to loop_counter2: sw x6, (&loop_counter2)
+ * - Return the values over UART.
+ * Faults are injected during the trigger_high & trigger_low.
+ * It needs to be ensured that the compiler does not optimize this code.
  *
  * @param uj The received uJSON data.
  */
-status_t handle_ibex_fi_init_trigger(ujson_t *uj) {
-  sca_select_trigger_type(kScaTriggerTypeSw);
-  // As we are using the software defined trigger, the first argument of
-  // sca_init is not needed. kScaTriggerSourceAes is selected as a placeholder.
-  sca_init(kScaTriggerSourceAes, kScaPeripheralIoDiv4);
+status_t handle_ibex_fi_char_mem_op_loop(ujson_t *uj) {
+  // Configure Ibex to allow reading ERR_STATUS register.
+  dif_rv_core_ibex_t rv_core_ibex;
+  if (dif_rv_core_ibex_init(
+          mmio_region_from_addr(TOP_EARLGREY_RV_CORE_IBEX_CFG_BASE_ADDR),
+          &rv_core_ibex) != kDifOk) {
+    return ABORTED();
+  }
```

Comment:
Good idea, I've create a function for that.

Reply:
We just discussed this in person. What you have right now makes sense because:
- We just read a single status register and feed that back to the host together with the result of the operation.
- The result of the operation may strongly depend on the specific test and we don't want to make the json more complicated.

Once we start adding more status register reads (e.g. alert handler status) it will make sense to encapsulate the status setup (now just `dif_rv_core_ibex_init()`) and the status reading into functions to reduce duplication of code and maintenance effort.